File: infra-market-quarkus/src/main/kotlin/io/infra/qk/repository/RolePermissionRepository.kt
package io.infra.qk.repository

import io.infra.qk.entity.RolePermission
import io.quarkus.hibernate.orm.panache.kotlin.PanacheRepository
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject

/**
 * 角色权限关联 Repository
 * @author liuqinglin
 * Date: 2025/8/30
 */
@ApplicationScoped
class RolePermissionRepository @Inject constructor() : PanacheRepository<RolePermission> {
    
    /**
     * 根据角色ID查找权限关联
     */
    fun findByRoleId(roleId: Long): List<RolePermission> {
        return find("roleId", roleId).list()
    }
    
    /**
     * 根据权限ID查找角色关联
     */
    fun findByPermissionId(permissionId: Long): List<RolePermission> {
        return find("permissionId", permissionId).list()
    }
    
    /**
     * 根据多个角色ID查找权限关联
     */
    fun findByRoleIds(roleIds: List<Long>): List<RolePermission> {
        if (roleIds.isEmpty()) {
            return emptyList()
        }
        return find("roleId in ?1", roleIds).list()
    }
    
    /**
     * 删除角色的所有权限关联
     */
    fun deleteByRoleId(roleId: Long) {
        delete("roleId", roleId)
    }
    
    /**
     * 删除权限的所有角色关联
     */
    fun deleteByPermissionId(permissionId: Long) {
        delete("permissionId", permissionId)
    }
}

File: infra-market-server-go/internal/repository/role_permission_repository.go
package repository

import (
	"github.com/bucketheadv/infra-market/internal/entity"
	"gorm.io/gorm"
)

type RolePermissionRepository struct {
	db *gorm.DB
}

func NewRolePermissionRepository(db *gorm.DB) *RolePermissionRepository {
	return &RolePermissionRepository{db: db}
}

// FindByRoleID 根据角色ID查询
func (r *RolePermissionRepository) FindByRoleID(roleID uint64) ([]entity.RolePermission, error) {
	var rolePermissions []entity.RolePermission
	err := r.db.Where("role_id = ?", roleID).Find(&rolePermissions).Error
	return rolePermissions, err
}

// FindByRoleIDs 批量查询
func (r *RolePermissionRepository) FindByRoleIDs(roleIDs []uint64) ([]entity.RolePermission, error) {
	if len(roleIDs) == 0 {
		return []entity.RolePermission{}, nil
	}
	var rolePermissions []entity.RolePermission
	err := r.db.Where("role_id IN ?", roleIDs).Find(&rolePermissions).Error
	return rolePermissions, err
}

// DeleteByRoleID 删除角色的所有权限关联
func (r *RolePermissionRepository) DeleteByRoleID(roleID uint64) error {
	return r.db.Where("role_id = ?", roleID).Delete(&entity.RolePermission{}).Error
}

// Create 创建角色权限关联
func (r *RolePermissionRepository) Create(rolePermission *entity.RolePermission) error {
	return r.db.Create(rolePermission).Error
}

// CountByPermissionID 统计权限的角色数量
func (r *RolePermissionRepository) CountByPermissionID(permissionID uint64) (int64, error) {
	var count int64
	err := r.db.Model(&entity.RolePermission{}).Where("permission_id = ?", permissionID).Count(&count).Error
	return count, err
}
